jkq/ddsim/pathqasmsimulator.py

```diff
@@ -0,0 +1,375 @@
+
+            # add the remaining matrix-vector multiplication
+            path.append((0, next_id))
+
+            # set contraction path
+            sim.set_contraction_path(path, True)
+
+        # determine the avg case contraction path for verification purpose
+        if mode == 'avgcase':
+            nGatesQC1 = options.get('nGatesQC1', 0)
+            nGatesQC2 = options.get('nGatesQC2', 0)
+            # print(nGatesQC1, nGatesQC2)
+            leftId = nGatesQC1
+            rightId = nGatesQC1 + 1
+            runId = nGatesQC1 + nGatesQC2 + 1
+            path = []
+            print(qobj_experiment.instructions[leftId].name)
+            # print(leftId, rightId, runId)
+            # todo create contraction path, determine right hand side strategie
+            for i in range(nGatesQC1):
+                if i == 0:
+                    path.append((leftId, rightId))
+                else:
+                    leftId = leftId - 1
+                    path.append((leftId, runId))
+                    runId = runId + 1
+                    if leftId == 1:
+                        leftId = leftId - 1
+                        path.append((leftId, runId))
+                        runId = runId + 1
+                        while rightId != nGatesQC1 + nGatesQC2:
+                            rightId = rightId + 1
+                            # print(runId, rightId)
+                            path.append((runId, rightId))
+                            runId = runId + 1
+                            if rightId == nGatesQC1 + nGatesQC2:
+                                break
+                    # print("step left", leftId, runId)
+                    if leftId == 0 and rightId == nGatesQC1 + nGatesQC2:
+                        # print("end values reached, breaking out")
+                        break
+                    for j in range(1, 3, 1):
+                        rightId = rightId + 1
+                        path.append((runId, rightId))
+                        runId = runId + 1
+                        # print("step right", leftId, rightId, runId)
+                        if rightId == nGatesQC1 + nGatesQC2:
+                            while leftId != 0:
+                                leftId = leftId - 1
+                                path.append((leftId, runId))
+                                runId = runId + 1
+                                # print("step left", leftId, rightId, runId)
+                                if leftId == 0:
+                                    # print("breaking out")
+                                    break
+                        if leftId == 0:
+                            # print("breaking out once more")
+                            break
+                if leftId == 0:
+                    # print("breaking out once more")
+                    break
+            # print(path)
+            # print(nGatesQC1, nGatesQC2)
+            sim.set_contraction_path(path, True)
```

Comment:
As stated in the earlier comments I would probably remove the `avgcase` for now, since it is not fully evaluated.